### PR TITLE
Implementing Camera Relative Ray Tracing

### PR DESCRIPTION
**Pull request by anisunity**
This PR takes advantage of the new API that allows us to do ray tracing in camera relative space. In the current state all tests are passing with camera relative rendering and without. Some bugs were found and solved during the process (filtering of reflections and point/directional shadows) which increase visual quality.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/PathTracing/Shaders/PathTracingLight.hlsl [hunk 1]
```diff
@@ -27,7 +27,11 @@ struct LightList
 
 bool IsRectAreaLightActive(LightData lightData, float3 position, float3 normal)
 {
+#if (SHADEROPTIONS_CAMERA_RELATIVE_RENDERING == 0)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/PathTracing/Shaders/PathTracingLight.hlsl:30]: This is not the correct way to fix this. Going back to where "position" come from it appear it come from

#ifdef HAS_LIGHTLOOP

    // Let's compute the world space position (the non-camera relative one if camera relative rendering is enabled)
    float3 shadingPosition = GetAbsolutePositionWS(fragInput.positionRWS);

mean that if you remove "GetAbsolutePositionWS" here. It "should" work. and 
IsRectAreaLightActive should only be: float3 lightVec = position - lightData.positionRWS;

Can check woth Manu for this one

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/PathTracing/Shaders/PathTracingLight.hlsl:30]: same for other in the file

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/PathTracing/Shaders/PathTracingLight.hlsl:30]: General reules is wherever we need 'GetAbsolutePositionWS', it mean there is certainly a problem somewhere

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/PathTracing/Shaders/PathTracingMain.raytrace [hunk 2]
```diff
@@ -79,7 +79,11 @@ void RayGen()
     {
         // Compute the ray direction from those coordinates (fast path for zero aperture)
         directionWS = -normalize(mul(jitteredPixelCoord, (float3x3)_PixelCoordToViewDirWS));
+    #if (SHADEROPTIONS_CAMERA_RELATIVE_RENDERING == 0)
         cameraPosWS = _WorldSpaceCameraPos;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/PathTracing/Shaders/PathTracingMain.raytrace:83]: We should use:

float3 GetPrimaryCameraPosition()
{
#if (SHADEROPTIONS_CAMERA_RELATIVE_RENDERING != 0)
    return float3(0, 0, 0);
#else
    return _WorldSpaceCameraPos;
#endif
}

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/PathTracing/Shaders/PathTracingMain.raytrace [hunk 3]
```diff
@@ -95,8 +99,14 @@ void RayGen()
 
         // Compute the new ray origin ( _ViewMatrix[0] = right, _ViewMatrix[1] = up, _ViewMatrix[2] = forward )
         float focusDistance = _PathTracedDoFConstants.y;
+    #if (SHADEROPTIONS_CAMERA_RELATIVE_RENDERING == 0)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/PathTracing/Shaders/PathTracingMain.raytrace:102]: We should create a function for this one

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/HDRaytracingLightCluster.cs [hunk 4]
```diff
@@ -336,19 +338,24 @@ void BuildGPULightVolumes(HDRayTracingLights rayTracingLights)
             for (int lightIdx = 0; lightIdx < rayTracingLights.reflectionProbeArray.Count; ++lightIdx)
             {
                 HDProbe currentEnvLight = rayTracingLights.reflectionProbeArray[lightIdx];
+
+                // Compute the camera relative position
+                Vector3 probePositionRWS = currentEnvLight.influenceToWorld.GetColumn(3);
+                probePositionRWS -= hdCamera.camera.transform.position;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/HDRaytracingLightCluster.cs:344]: do this only if camera relative is enabled

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/HDRaytracingLightCluster.cs:344]: like in lightloop.cs

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/HDRaytracingLightCluster.cs [hunk 5]
```diff
@@ -293,14 +293,16 @@ void BuildGPULightVolumes(HDRayTracingLights rayTracingLights)
                     // Reserve space in the cookie atlas
                     m_RenderPipeline.ReserveCookieAtlasTexture(currentLight, light, currentLight.type);
 
-               
+                    // Compute the camera relative position
+                    Vector3 lightPositionRWS = currentLight.gameObject.transform.position - hdCamera.camera.transform.position;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/HDRaytracingLightCluster.cs:297]: there should be a if here like iin lightloop.cs, we do the relative only if camera relative is enabled, so the shader  code don't need to have lot of ifdef with GetAbsolutePosition

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/HDRaytracingLightCluster.cs [hunk 6]
```diff
@@ -367,7 +374,7 @@ void EvaluateClusterVolume(HDCamera hdCamera)
         {
             var settings = hdCamera.volumeStack.GetComponent<LightCluster>();
 
-            clusterCenter = hdCamera.camera.gameObject.transform.position;
+            clusterCenter.Set(0, 0, 0);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/HDRaytracingLightCluster.cs:377]: same, test if camera relative

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Common/AtmosphericScatteringRayTracing.hlsl [hunk 7]
```diff
@@ -8,7 +8,11 @@ void ApplyFogAttenuation(float3 origin, float3 direction, float t, inout float3
     if (_FogEnabled)
     {
         float dist = min(t, _MaxFogDistance);
+#if (SHADEROPTIONS_CAMERA_RELATIVE_RENDERING == 0)
         float absFogBaseHeight = GetAbsolutePositionWS(float3(0.0, _HeightFogBaseHeight, 0.0)).y;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Common/AtmosphericScatteringRayTracing.hlsl:12]: _HeightFogBaseHeight should also be handle on C# side like we do for rasterizatin:

            float crBaseHeight = baseHeight.value;

            if (ShaderConfig.s_CameraRelativeRendering != 0)
            {
                crBaseHeight -= hdCamera.camera.transform.position.y;
            }

            float layerDepth = Mathf.Max(0.01f, maximumHeight.value - baseHeight.value);
            float H = ScaleHeightFromLayerDepth(layerDepth);
            cb._HeightFogExponents = new Vector2(1.0f / H, H);
            cb._HeightFogBaseHeight = crBaseHeight;

Comment (sebastienlagarde) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Common/AtmosphericScatteringRayTracing.hlsl:12]: currently it is already handle, this _HeightFogBaseHeight  should be share with raytrace part? Or maybe we wait for the PR of Julien?

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/DebugLightCluster.compute [hunk 8]
```diff
@@ -33,8 +33,10 @@ void DebugLightCluster(uint2 groupThreadId : SV_GroupThreadID, uint2 groupId : S
     }
 
     // Compute the real world space position of this pixel
-    PositionInputs posInput = GetPositionInput(currentPixelCoordinate, 1.0/ _ScreenSize.xy, depth, UNITY_MATRIX_I_VP, GetWorldToViewMatrix(), 0);
+    PositionInputs posInput = GetPositionInput(currentPixelCoordinate, 1.0 / _ScreenSize.xy, depth, UNITY_MATRIX_I_VP, GetWorldToViewMatrix(), 0);
+#if (SHADEROPTIONS_CAMERA_RELATIVE_RENDERING == 0)
     posInput.positionWS = GetAbsolutePositionWS(posInput.positionWS);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/DebugLightCluster.compute:38]: hmm... this is not expected to have GetAbsolutePositionWS here

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Common/AtmosphericScatteringRayTracing.hlsl (hunk 7 above)
Comment (anisunity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Raytracing/Shaders/Common/AtmosphericScatteringRayTracing.hlsl]: corrected

**Comment by anisunity** [comment]
This PR is 2020.2 only

**Review by sebastienlagarde — APPROVED** [review]

**Review by sebastienlagarde — APPROVED** [review]

**Pull request merged**